README.md
<a id="intro"></a>

### Introduction

CloudVPS Boss is a backup suite for Linux that utilises Duplicity to backup and restore files, folders and databases from and to Openstack Swift with optional encryption, in our case the CloudVPS Object Store.

Backing up to this medium means that your data is stored three times over two different datacenters for a very low price. The name means “CloudVPS Backup to Object Store Script” and was chosen by our marketing team (but the development team obviously loves it).

The code is open source so you can use it on non CloudVPS object stores in the unlikely case that you would want to do that.

### Table of Contents

<ol>
<li><a href="#intro">Introduction</a></li>
<li><a href="#req">Requirements</a></li>
<li><a href="#ins">Installation</a></li>
  <ul>
  <li><a href="#mai">Manual Installation</a></li>
  <li><a href="#ce5">Notes on CentOS</a></li>
  <li><a href="#arc">A note on Arch Linux</a></li>
  </ul>
<li><a href="#con">Configuration</a></li>
<ul>
  <li><a href="#customconfig"></a>Custom Configuration</li>
  <li><a href="#bwlimit"></a>Bandwidth limits</li>
</ul>
<li><a href="#bac">Backup</a></li>
<ul>
  <li><a href="#dat">A note on database dumps</a></li>
  <li><a href="#progress">Backup Progress</a></li>
</ul>
<li><a href="#restore">Restore</a></li>
  <ul>
  <li><a href="#encrestore1">Restoring encrypted backups</a></li>
  <li><a href="#full">How do I restore a full backup?</a></li>
  <li><a href="#whe">Where are my files/databases restored?</a></li>
  <li><a href="#oth">Recovery from another host</a></li>
  <li><a href="#err">Errors during restore</a></li>
  <li><a href="#man">Manual Restore</a></li>
  <li><a href="#sqr">A note database restore</a></li>
  </ul>
<li><a href="#inf">Information and Statistics</a></li>
  <ul>
    <li><a href="#viewfiles">How can I view the full contents of a backup?</a></li>
    <li><a href="#manualclean">Manually clean up / remove backups</a></li>
  </ul>
<li><a href="#not">Notifications and Monitoring</a></li>
<li><a href="#uni">Uninstall</a></li>
<li><a href="#enc">Duplicity Encryption</a></li>
    <ul><li><a href="#encwhy">Why Encryption</a></li>
    <li><a href="#encsetup">Set up encryption</a></li>
    <li><a href="#encrestore">Restore encryption keys and config</a></li></ul>
<li><a href="#oth">Other Edge Cases</a></li>
  <ul><li><a href="#del">What if my Container is deleted?</a></li>
  <li><a href="#pass">What if my Password Changed?</a></li>
  <li><a href="#dow">What if the Object Store is Down?</a></li>
  <li><a href="#run">What happens if a backup is still running when a new one starts?</a></li>
  <li><a href="#free">Not enough free space</a></li>
  <li><a href="#scheme">Unsupported Backup Scheme</a></li>
  <li><a href="#logger">logger: command not found</a></li>
  <li><a href="#performance">Duplicity is using 100% of all resources</a></li>
  <li><a href="#hugefiles">Very large backups</a></li>
  <li><a href="#screen">Restore looks 'weird' in a screen session</a></li>
  </ul>
<li><a href="#lic">License</a></li>
</ol>

<a id="req"></a>

### Requirements

CloudVPS Boss is tested and should work on and with the following distributions and software:

- Ubuntu 12.04 and up
- Debian 7, 8
- CentOS 6, 7

- MySQL/MariaDB 5.0 - 5.6
- PostgreSQL 9.0 - 9.4

The following distributions are NOT supported. That means, if something breaks, you're on your own:

- Ubuntu < 12.04
- Debian 6
- CentOS 5

<a id="ins"></a>

### Installation

#### Object Store user

Make sure you have an Openstack or Object Store project. If not, please contact sales or order one via the website.

The software needs a tenant ID, username and password of an Object Store account. A tenant ID is the same as the Project ID, not the user ID.

If you back up multiple servers to the same container, they are able to access all the backups. If you do not want that machines can see each others backup you need to create a seperate project and user in that project per server you want to backup. Or, you can setup encryption of the backups, see the "Encryption" section for more info on that.

Please do not use your main account, but create a new user under your projects:

- Login to the interface over at https://interface.cloudvps.com.

- Navigate to "User Manager" --> "User Manager"

- Click the "+" and fill in the required data. Note down the password, you will need it later. The image below is an example, please use a meaningfull username like `backups@mycompany.nl`.

![](http://i.imgur.com/rxVnCd6.png)

- Add the user to your openstack project. Click the "+" and select the correct project. Give it only "swiftoperator" permissions.
.

- Note down the Project ID. This is the same as the Tenant ID.

This account will be used to place the backup in the object store. If you want, you can create an account for every server you want to backup with different credentials. Make sure the username is descriptive.

The object store account can already contain data. The backups are written to the container `cloudvps-boss-backup`. This container is allowed to have pre-existing data, but it is not recommended. The container name cannot be changed. Multiple servers can be backed up to the container, as long as the hostname configured is different.

#### Installation

You need to run the installer as the `root` user, or as a user with an `EUID` equal to 0. All the cloudvps-boss tools will check if they are executed as root and will fail otherwise.

Download the .tar.gz file containing the latest version of the script:

    wget -O cloudvps-boss-latest.tar.gz https://download.cloudvps.com/cloudvps-boss/cloudvps-boss_latest.tar.gz

Extract it:

    tar -xf cloudvps-boss_latest.tar.gz

Go in the folder and start the installer:

    cd cloudvps-boss
    bash install.sh

If you want to start the installation unattended, give the username and password as options:

    cd cloudvps-boss
    bash install.sh "username@example.org "p@ssw0rd" "tenant id"

A one liner, if you need to do multiple installations:

    rm -rf /tmp/cloudvps*; wget -O /tmp/cloudvps-boss_latest.tar.gz https://download.cloudvps.com/cloudvps-boss/cloudvps-boss_latest.tar.gz && cd /tmp && tar -xf cloudvps-boss_latest.tar.gz && cd /tmp/cloudvps-boss && bash install.sh

The installer will ask you for your Openstack Username, Password and Tenant ID. Your pasword will not be shown when you type it, you will also not see stars or some other masking.

A version of Duplicity with support for OpenStack Swift will be installed, as well as the required build tools, and the necessary openstack clients.

If the csf or lfd firewall applications are detected, IP ranges related to the CloudVPS Object Store will be configured as exception rules.

After that it will create or update the container `cloudvps-boss-backup`. It will also place the required config files and such.

Example install output:

    # CloudVPS Boss Install 1.6 started on Sun Aug 24 08:56:14 CEST 2014.
    # Adding exceptions for lfd.
    # Adding exceptions for csf.
    # Backing up /etc/cloudvps-boss to /var/backups/cloudvps-boss.17118
    # Starting install_duplicity.sh
    # CloudVPS Boss Duplicity Install 0.6.24 started on Sun Jul 27 10:35:04 CEST 2014.
    # Ubuntu 12.04
    /usr/local/cloudvps-boss/duplicity/duplicity-0.6.24 /tmp
    /tmp
    # Duplicity installation script completed
    # install_duplicity.sh completed.
    # Starting credentials.sh
    # CloudVPS Boss Credentials Config 1.6 started on Sun Aug 24 08:56:14 CEST 2014.

    Openstack Username (user@example.org): user@example.org
    Openstack Password (not shown):
    Openstack Tenant ID : 22[...]8a

    # Username: Example Tenant:user@example.org
    # Auth URL: https://identity.stack.cloudvps.com/v2.0
    # Written auth config to /etc/cloudvps-boss/auth.conf.
    # Checking Swift Container for Backups: https://public.objectstore.eu/v1/22[...]8a/cloudvps-boss-backup/
    # credentials.sh completed.
    # Randomized cronjob time, will run on 4:36.
    # Creating symlink for backup script in /usr/local/bin/cloudvps-boss.
    # Creating symlink for recovery script in /usr/local/bin/cloudvps-boss-restore.
    # Creating symlink for status script in /usr/local/bin/cloudvps-boss-stats.
    # Creating symlink for update script in /usr/local/bin/cloudvps-boss-update.

    # CloudVPS Boss installation completed.


If you upgrade the script or by accident run the installer again it will backup all the config it can find and then place new ones. It will backup and place back the auth, backup and exclude config. Updating can be done at any time with the following command:

    cloudvps-boss-update

The cronjob placed will auto update the script every month on the first day of the month.

<a id="mai"></a>

#### Manual Installation

If you are on a unsupported distribution you need to to the following things to get cloudvps-boss working.

- Install python 2.7.
- Install python-swiftclient and python-keystoneclient.
- Make sure wget, curl, openssl, awk, which, sed and grep are available.
- If you want encryption, also install gpg.
- Install/Compile Duplicity, a version higher than 0.6.22.
- Create the folder /etc/cloudvps-boss
- Create the pre-backup.d, post-fail-backup.d and post-backup.d folders in there
- Place the required files there, like so:

        /etc/cloudvps-boss/
        |-- auth.conf
        |-- backup.conf
        |-- cloudvps-boss-encryption-setup.sh
        |-- cloudvps-boss-restore.sh
        |-- cloudvps-boss-stats.sh
        |-- cloudvps-boss-list-current-files.sh
        |-- cloudvps-boss-update.sh
        |-- cloudvps-boss.sh
        |-- common.sh
        |-- exclude.conf
        |-- post-backup.d
        |   `-- 10-upload-completed-status.sh
        |-- post-fail-backup.d
        |   |-- 10-upload-fail-status.sh
        |   `-- 20-failure-notify.sh
        |-- pre-backup.d
        |   |-- 10-upload-starting-status.sh
        |   |-- 15-mysql_backup.sh
        |   `-- 15-postgresql_backup.sh
        `-- uninstall.sh


- Create symlinks in /usr/local/bin to cloudvps-boss, cloudvps-boss-restore, cloudvps-boss-update, cloudvps-list-current-files and cloudvps-boss-stats.
- Place the cloudvps-boss.cron file in /etc/cron.d/
- Configure auth.conf and backup.conf. (See below, section Configuration)
- Optional, set up encryption.

<a id="ce5"></a>

#### Notes on CentOS

On CentOS 5 we also compile and install a newer version of sqlite, setuptools and Python version 2.7. We don't touch the system Python 2.4, so Yum still works.

Another step we take on this distro, if needed, is updating the system root CA certicicates file, making a safety copy of the original. This is necessary for the swift client, to allow it to verify the CloudVPS SSL credentials correctly.

For all CentOS releases, we disable all the repositories except `base` during installation. The script installs the `epel` repository and uses this for the python packages. If you run RHEL, Scientific Linux or Oracle Linux, make sure you either enable or rename the repository which contains base packages to `base` otherwise the installation will fail.

<a id="arc"></a>

#### Notes Arch Linux

Arch Linux is not supported, but we do provide a (as of August 2014) working installer. It however fails if you have gcc-multilib installed. Install the required packages manually then.

<a id="con"></a>

### Configuration

The install script creates a default configuration for the backup process, with these settings in it:

- Create a full backup every 14 days
- Keep at max 6 full backups
- Create an incremental backup on the other days, once a day.

This gives you 3 months of retention. The settings can be found in `/etc/cloudvps-boss/backup.conf`:

    # Server hostname. Will be replaced during install. Must be unique among backuped servers.
    HOSTNAME="replace_me"
    # Create a full backup if the last is older than 14 days.
    FULL_IF_OLDER_THAN="14D"
    # Keep at max 6 full backups.
    FULL_TO_KEEP="6"
    # Only change this if your tmp folder is to small. See README
    TEMPDIR="/tmp"

If you want more or less retention, change the variables. For one week of retention, create a full backup if the other full is older than 7 days and keep at max 1 full backup. If you want a month of retention, create a full backup if the other full is older than 7 days and keep at max 4 full backups.

You can increase the amount of days before a full backup is created. Duplicity will create more incremental backups then.

Note that full backups have a positive effect on restore speeds, but a negative effect on the amount of storage needed. Increase the relative number of full backups for better speed, or lower it for less storage.

To get 6 months of retention with just one full backup, create a full backup if the other full is older than 6 months (6M) and keep at max 1 full backup.

The auth.conf file has the credentials needed for Swift and Openstack authentication:

    SWIFT_USERNAME="TENANT-ID:USERNAME"
    SWIFT_PASSWORD="super_secure_password"
    SWIFT_AUTHURL="https://identity.stack.cloudvps.com/v2.0"
    SWIFT_AUTHVERSION="2"
    OS_AUTH_URL="https://identity.stack.cloudvps.com/v2.0"
    OS_TENANT_NAME="TENANT-ID"
    OS_USERNAME="USERNAME"
    OS_PASSWORD="super_secure_password"
    OS_TENANT_ID="TENANT-ID"

Make sure the tenant has access to swift, the swiftoperator role. Compute is not required. The tenant id is the same as the project id.

If a backup runs longer than 24 hours, and if an email addresses and mailing is configured, you will receive an email about this. If backups regularly take longer than 24 hours you might need to investigate why that is the case.

<a id="customconfig"></a>

#### Custom Configuration

If you want to use a custom backend or extra Duplicity options you can configure those in the following file:

    /etc/cloudvps-boss/custom.conf

This file is not there by default, you should create it.

The following options are supported:

    BACKUP_BACKEND=''
    CONTAINER_NAME=''
    CUSTOM_DUPLICITY_OPTIONS=''

If you want to use a container other than the default 'cloudvps-boss-backup' you can change CONTAINER_NAME.

You can specify a custom backend here, this will be used instead of our Object Store. You can also specify custom options which will be added to every Duplicity command. For example, to use an FTP server:

    BACKUP_BACKEND='ftp://backup@example.com:Passw0rd@example.com'

For FTP you need to install `ncftp`.

To use an SSH/SFTP server with a custom SSH key:


CHANGELOG.md
# Changelog

## 1.9.17

- Update Duplicity to 0.7.17

## 1.9.16

- Add support for Debian 9

## 1.9.16

- Update Duplicity to 0.7.16

## 1.9.13

- Update Duplicity to 0.7.15
- Add manual full backup command

## 1.9.12

- Update Duplicity to 0.7.14
- Remove pygobj dependency
- Don't chmod backup folder
- Remove duplicate log from error email
- Add new object store IP range to firewall

## 1.9.11

- Update Duplicity to 0.7.13
- Remove progress script

## 1.9.10

- Update Duplicity to 0.7.10
- Do not run fail scripts after a failed cleanup.

## 1.9.9

- Update Duplicity to 0.7.09
- Fix typo's

## 1.9.8

- Update Duplicity to 0.7.08
- Remove Arch and Fedora from the installer
- Use pip requirements for install on CentOS 7 & Debian 7
- Fix issue in cleanup script


## 1.9.7

- Update Duplicity to 0.7.07.1
- Add support for Ubuntu 16.04
- Remove CentOS 5 and Debian 6 from the installer
- Fix issue in cleanup script

## 1.9.6

- Use pip requirements for install on Debian/CentOS 6
- Remove async upload
- Add apt-get update before first package install

## 1.9.5

- Add cleanup command, wrapper around duplicity cleanup.
- Install specific versions of python libraries on Debian 6. Newer won't work with python 2.6.
- Update uninstaller to remove more python libraries.

## 1.9.4

- Fix race condition

## 1.9.3

- Install specific versions of swiftclient, keystoneclient and oslo on CentOS 6. Newer won't work with python 2.6.

## 1.9.2

- Fix install issue with mixed pip/repo packages.

## 1.9.1

- Remove requests workaround for ubuntu 14.04.
- Increase retry time.
- Update Duplicity to 0.7.05

## 1.9.0

- Add package installation on install (curl was unavailable sometimes).
- Add suggestion in error messages on failed install to let user retry the command.
- Retry MySQL credential building when authentication fails, before emailing user.
- Improve lockfile / running backup check.
- Fix large (+250 GB if 25 MB volumes or 2.5 TB if 250 MB volumes) backup sets. Only full backups were made because of https://launchpad.net/+branch/~raymii/duplicity/fix-swiftbackend-max-10000-files-in-list.
- Update some duplicity command options because of deprecation.
- Upgrade Duplicity version to 0.7.04.

## 1.8

- Add send email if MySQL backup fails.
- Change default volsize to 250 MB.
- Upgrade Duplicity version to.0.7.03.
- Upgrade Python to 2.7.10 (on new installs)

## 1.7

- Add check during install/update for already compiled dependencies.
- Add install of base-devel to Arch installer.
- Add more information to failure email script.
- Add lock file checking and handling.
- Add default MySQL data dir to exclude list.
- Change installer download paths to more clean structure.
- Upgrade Duplicity version to 0.7.01.
- Upgrade Duplicity version to 0.7.02.
- Remove overwriting of cronjob during upgrade.
- Fix progress reporting error when df does not support --total.
- Fix rare AUTH_TOKEN not found during install.
- Fix curl dependency check in installer
- Fix curl not found error in credential script



## 1.6

- Add view progress command for long running backups.
- Add section for large backups to README.
- Add https download link to README.
- Add bandwidth limit instructions to README.
- Add wget/curl wrapper for remote file downloading.
- Add more clear encryption instructions.
- Add workaround when AUTH_TOKEN during credential setup fails.
- Change default exclude list.
- Change update download link to https.
- Change internal json functions to be more clear.
- Change default retention to 3 months instead of 6.
- Fix installer bug when multiple duplicity source folders exist.
- Fix installer bug when install fails on immutable scripts.
- Remove some non used code.

## 1.5

- Add custom backend support.
- Add custom additional options support.
- Add use of duplicity --name parameter.
- Add documentation on manual cleanup.
- Change auto update to run once a week.
- Fix uninstaller when pip was not named pip.
- Fix some quoting issues.

## 1.4

- Add encryption setup support.
- Add yum clean before update and install
- Add /home/virtfs to exclude list (cpanel)
- Add upload of version number to status folder during updates
- Change more clear documentation on projects and user accounts.
- Fix installation error on CentOS 5 sqlite
- Fix installation error on cpanel + Centos 6.
- Fix no compilation in /tmp (noexec errors)

## 1.3

- Initial public release


install.sh
#!/bin/bash
# CloudVPS Boss - Duplicity wrapper to back up to OpenStack Swift
# Copyright (C) 2018 Remy van Elst. (CloudVPS Backup to Object Store Script)
# Author: Remy van Elst, https://raymii.org
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#

set -o pipefail

VERSION="1.9.17"
TITLE="CloudVPS Boss Install ${VERSION}"

if [[ ${DEBUG} == "1" ]]; then
    set -x
fi

lecho() {
    logger -t "cloudvps-boss" -- "$1"
    echo "# $1"
}

lerror() {
    logger -t "cloudvps-boss" -- "ERROR - $1"
    echo "$1" 1>&2
}

log() {
    logger -t "cloudvps-boss" -- "$1"
}

if [[ "${EUID}" -ne 0 ]]; then
   lerror "This script must be run as root."
   exit 1
fi

lecho "${TITLE} started on $(date)."

usage() {
    echo "Usage:"
    echo "To install cloudvps-boss with interactive username password question:"
    echo "./$0"
    echo; echo "To install cloudvps-boss non-interactive:"
    echo "./$0 username@domain.tld 'passw0rd' 'tenant id'"
}

if [[ ! -z "$1" ]]; then
    if [[ "$1" == "help" ]]; then
        usage
    fi
fi

run_script() {
    # check if $1 is a file and execute it with bash.
    # log result and exit if script fails.
    if [[ -f "$1" ]]; then
        log "Starting $1"
        bash "$1" "$2" "$3" "$4"
        if [[ $? == 0 ]]; then
            logger -t "cloudvps-boss" -- "$1 completed."
        else
            lerror "$1 did not exit cleanly."
            exit 1
        fi
    else
        lerror "Cannot find $1."
    fi
}

command_exists() {
    # check if command exists and fai otherwise
    command -v "$1" >/dev/null 2>&1
    if [[ $? -ne 0 ]]; then
        lerror "I require $1 but it's not installed. Please install it. I've tried to install it but that failed. Aborting."
        exit 1
    fi
}

## Why all this effort to manually install stuff in three different places you wonder? Because some people refuse to build images with curl included. Why? Who knows... That explains these next 100 or so lines of extra code...

distro_version() {
    if [[ -f "/etc/debian_version" ]]; then
        NAME="Debian"
        VERSION="$(awk -F. '{print $1}' /etc/debian_version)"
    fi
    if [[ -f "/etc/lsb-release" ]]; then
        NAME="$(awk -F= '/DISTRIB_ID/ {print $2}' /etc/lsb-release)"
        VERSION="$(awk -F= '/DISTRIB_RELEASE/ {print $2}' /etc/lsb-release)"
    fi
    if [[ -f "/etc/redhat-release" ]]; then
        NAME="$(awk '{ print $1 }' /etc/redhat-release)"
        VERSION="$(grep -Eo "[0-9]\.[0-9]" /etc/redhat-release | cut -d . -f 1)"
    fi
    if [[ -f "/etc/arch-release" ]]; then
        NAME="Arch"
        VERSION="Rolling"
    fi
    if [[ "$1" == "name" ]]; then
        echo "${NAME}"
    fi
    if [[ "$1" == "version" ]]; then
        echo "${VERSION}"
    fi
}

get_hostname() {
    # Try Openstack Metadata service first
    HOSTNAME="$(curl -m 3 -s http://169.254.169.254/openstack/latest/meta_data.json | grep -o '\"uuid\": \"[^\"]*\"' | awk -F\" '{print $4}')"
    if [[ -z "${HOSTNAME}" ]]; then
        # Otherwise XLS /var/fistboot
        if [[ -f "/var/firstboot/settings" ]]; then
            HOSTNAME="$(awk -F= '/hostname/ {print $2}' /var/firstboot/settings)"
        else
            # ask the system if all else fails.
            HOSTNAME="$(uname -n)"
        fi
    fi

    echo "${HOSTNAME}"
}

install_packages_debian() {
    lecho "Installing packages required for installation."
    APT_UPDATE="$(apt-get -qq -y --force-yes update > /dev/null 2>&1)"
    if [[ "$?" -ne 0 ]]; then
        lerror "'apt-get update' failed."
        exit 1
    fi
    for PACKAGE in awk sed grep tar gzip which openssl curl wget screen vim haveged unattended-upgrades; do
        /usr/bin/apt-get -qq -y --force-yes -o "Dpkg::Options::=--force-confdef" -o "Dpkg::Options::=--force-confold" install "${PACKAGE}" >/dev/null 2>&1
    done
}

install_packages_centos() {
    lecho "Installing packages required for installation."
    for PACKAGE in awk sed grep tar gzip which openssl curl wget screen vim haveged yum-cron; do
        yum -q -y --disablerepo="*" --disableexcludes=main --enablerepo="base" --enablerepo="updates" install "${PACKAGE}" >/dev/null 2>&1
    done
}

DISTRO_NAME=$(distro_version name)
DISTRO_VERSION=$(distro_version version)

case "${DISTRO_NAME}" in
    "Debian")
        lecho "Debian ${DISTRO_VERSION}"
        install_packages_debian
    ;;
    "Ubuntu")
        lecho "Ubuntu ${DISTRO_VERSION}"
        install_packages_debian
    ;;
    "CentOS")
        lecho "CentOS ${DISTRO_VERSION}"
        install_packages_centos
    ;;

    *)
    lerror "Distro unknown or not supported"
    exit 1
    ;;
esac

for COMMAND in "awk" "sed" "grep" "tar" "gzip" "which" "openssl" "curl"; do
    command_exists "${COMMAND}"
done

if [[ -f "/etc/csf/csf.fignore" ]]; then
    # Add ourself to the csf file ignore list
    # lfd will not scan and mark us suspicious
    if [[ ! "$(grep 'cloudvps-boss' /etc/csf/csf.fignore)" ]]; then
        lecho "Adding exceptions for lfd."
        echo "/tmp/pip-build-root/*" >> /etc/csf/csf.fignore
        echo "/tmp/cloudvps-boss/*" >> /etc/csf/csf.fignore
        echo "/usr/local/cloudvps-boss/*" >> /etc/csf/csf.fignore
        echo "/etc/cloudvps-boss/*" >> /etc/csf/csf.fignore
        service lfd restart 2>&1 > /dev/null
    fi
    if [[ ! "$(grep '89.31.101.64' /etc/csf/csf.allow)" ]]; then
        # for regular iptables
        #-A OUTPUT -d 89.31.101.64/27 ! -o lo -p tcp -m tcp --dport 443 -j ACCEPT
        # Add a rule for the CloudVPS object store to CSF
        lecho "Adding exceptions for csf."
        csf -a "tcp|out|d=443|d=89.31.101.64/27" "CloudVPS Boss Object Store for backup" 2>&1 > /dev/null
        service csf restart 2>&1 > /dev/null
        csf -r 2>&1 > /dev/null
    fi
    if [[ ! "$(grep '31.3.100.121' /etc/csf/csf.allow)" ]]; then
        # for regular iptables
        #-A OUTPUT -d 31.3.100.121/29 ! -o lo -p tcp -m tcp --dport 443 -j ACCEPT
        # Add a rule for the CloudVPS object store to CSF
        lecho "Adding exceptions for csf."
        csf -a "tcp|out|d=443|d=31.3.100.121/29" "CloudVPS Boss Object Store for backup 2" 2>&1 > /dev/null
        service csf restart 2>&1 > /dev/null
        csf -r 2>&1 > /dev/null
    fi
fi

if [[ -d "/etc/cloudvps-boss" ]]; then
    # check if we already exist, if so, back us up
    lecho "Backing up /etc/cloudvps-boss to /var/backups/cloudvps-boss.$$"
    if [[ ! -d "/var/backups/cloudvps-boss.$$" ]]; then
        mkdir -p "/var/backups/cloudvps-boss.$$"
        if [[ "$?" -ne 0 ]]; then
            lerror "Cannot create folder /var/backups/cloudvps-boss.$$"
        fi
    fi

    cp -r "/etc/cloudvps-boss" "/var/backups/cloudvps-boss.$$"
    if [[ "$?" -ne 0 ]]; then
        lerror "Cannot backup /etc/cloudvps-boss to /var/backups/cloudvps-boss.$$."
        exit 1
    fi

    if [[ -f "/etc/cron.d/cloudvps-boss" ]]; then
        cp -r "/etc/cron.d/cloudvps-boss" "/var/backups/cloudvps-boss.$$/cloudvps-boss.cron.bak"
        if [[ "$?" -ne 0 ]]; then
            lerror "Cannot backup /etc/cron.d/cloudvps-boss to /var/backups/cloudvps-boss.$$/cloudvps-boss.cron.bak."
            exit 1
        fi
    fi
fi

for FOLDER in "/etc/cloudvps-boss/pre-backup.d" "/etc/cloudvps-boss/post-backup.d" "/etc/cloudvps-boss/post-fail-backup.d"; do
    # create a few required folders
    if [[ ! -d "${FOLDER}" ]]; then
        mkdir -p "${FOLDER}"
        if [[ $? -ne 0 ]]; then
        lerror "Cannot create ${FOLDER}"
            exit 1
        fi
    fi
done

log "Extracting to /etc/cloudvps-boss/"
# we copy all the things manually because
# some users do a chattr +i on stuff they don't want
# overwritten. A cp -r fails and leaves inconsistent state,
# a manual copy only fails the chattr'd things.
for COPY_FILE in "README.md" "LICENSE.md" "CHANGELOG.md"; do
    cp "${COPY_FILE}" "/etc/cloudvps-boss/${COPY_FILE}"
    if [[ "$?" -ne 0 ]]; then
        lerror "Cannot copy ${COPY_FILE} to /etc/cloudvps-boss/${COPY_FILE}."
    fi
done

for COPY_FILE in "cloudvps-boss.cron" "backup.conf" "cloudvps-boss-encryption-setup.sh" "cloudvps-boss-list-current-files.sh" "cloudvps-boss-verify.sh" "cloudvps-boss-cleanup.sh" "cloudvps-boss-restore.sh" "cloudvps-boss.sh" "cloudvps-boss-stats.sh" "cloudvps-boss-manual-full.sh" "cloudvps-boss-update.sh" "common.sh" "exclude.conf" "uninstall.sh"; do
    cp "cloudvps-boss/${COPY_FILE}" "/etc/cloudvps-boss/${COPY_FILE}"
    if [[ "$?" -ne 0 ]]; then
        lerror "Cannot copy cloudvps-boss/${COPY_FILE} to /etc/cloudvps-boss/${COPY_FILE}."
    fi
done

for COPY_FILE in "10-upload-starting-status.sh" "11_lockfile_check.sh" "15-mysql_backup.sh" "15-postgresql_backup.sh"; do
    cp "cloudvps-boss/pre-backup.d/${COPY_FILE}" "/etc/cloudvps-boss/pre-backup.d/${COPY_FILE}"
    if [[ "$?" -ne 0 ]]; then
        lerror "Cannot copy cloudvps-boss/${COPY_FILE} to /etc/cloudvps-boss/pre-backup.d/${COPY_FILE}."
    fi
done

for COPY_FILE in "10-upload-completed-status.sh"; do
    cp "cloudvps-boss/post-backup.d/${COPY_FILE}" "/etc/cloudvps-boss/post-backup.d/${COPY_FILE}"
    if [[ "$?" -ne 0 ]]; then
        lerror "Cannot copy cloudvps-boss/${COPY_FILE} to /etc/cloudvps-boss/post-backup.d/${COPY_FILE}."
    fi
done

for COPY_FILE in "10-upload-fail-status.sh" "20-failure-notify.sh"; do
    cp "cloudvps-boss/post-fail-backup.d/${COPY_FILE}" "/etc/cloudvps-boss/post-fail-backup.d/${COPY_FILE}"
    if [[ "$?" -ne 0 ]]; then
        lerror "Cannot copy cloudvps-boss/${COPY_FILE} to /etc/cloudvps-boss/post-fail-backup.d/${COPY_FILE}."
    fi
done

# See if we are upgrading and if so
# place back the important config files
for CONF_FILE in "auth.conf" "email.conf" "backup.conf" "custom.conf" "exclude.conf" "encryption.conf"; do
    if [[ -f "/var/backups/cloudvps-boss.$$/cloudvps-boss/${CONF_FILE}" ]]; then
        lecho "Update detected. Placing back file ${CONF_FILE}."
        cp -r "/var/backups/cloudvps-boss.$$/cloudvps-boss/${CONF_FILE}" "/etc/cloudvps-boss/${CONF_FILE}"
    fi
done

# complicated loop to run the installer and the credentials script
# with the correct parameters.
for SCRIPT in "install_duplicity.sh" "credentials.sh"; do
    if [[ "${SCRIPT}" == "credentials.sh" ]]; then
        if [[ ! -z "$1" ]]; then
            if [[ ! -z "$2" ]]; then
                if [[ ! -z "$3" ]]; then
                    run_script "${SCRIPT}" "$1" "$2" "$3"
                else
                   run_script "${SCRIPT}"
                fi
            else
                run_script "${SCRIPT}"
            fi
        else
           run_script "${SCRIPT}"
        fi
    else
        run_script "${SCRIPT}"
    fi
done

# hostname is used by Duplicity...
HOSTNAME="$(get_hostname)"
# get and set the hostname in the config. Fails if config is chattr +i.
sed -i "s/replace_me/${HOSTNAME}/g" /etc/cloudvps-boss/backup.conf

if [[ ! -d "/etc/cron.d" ]]; then
    mkdir -p "/etc/cron.d"
fi

if [[ ! -f "/etc/cron.d/cloudvps-boss" ]]; then
    mv "/etc/cloudvps-boss/cloudvps-boss.cron" "/etc/cron.d/cloudvps-boss"
    if [[ "$?" -ne 0 ]]; then
        lerror "Cannot place cronjob in /etc/cron.d."
    fi
    # use awk to get a number between 0 and 6 for the hour
    RANDH="$(awk 'BEGIN{srand();print int(rand()*(0-6))+6 }')"
    RANDM="$(awk 'BEGIN{srand();print int(rand()*(0-59))+59 }')"
    # and 0 to 59 for the minutes. Then place it in the cronjob.
    sed -i -e "s/RANDH/${RANDH}/g" -e "s/RANDM/${RANDM}/g" /etc/cron.d/cloudvps-boss
    # and show the user
    lecho "Randomized cronjob time, will run on ${RANDH}:${RANDM}."
fi

if [[ ! -d "/usr/local/bin" ]]; then
    mkdir -p "/usr/local/bin"
fi

for COMMAND in "cloudvps-boss.sh" "cloudvps-boss-restore.sh" "cloudvps-boss-stats.sh" "cloudvps-boss-update.sh" "cloudvps-boss-list-current-files.sh" "cloudvps-boss-manual-full.sh"; do
    log "Creating symlink for /etc/cloudvps-boss/${COMMAND} in /usr/local/bin/${COMMAND%.sh}."
    chmod +x "/etc/cloudvps-boss/${COMMAND}"
    ln -fs "/etc/cloudvps-boss/${COMMAND}" "/usr/local/bin/${COMMAND%.sh}"
done

for FILE in "pre-backup.d/15-mysql_backup.sh" "pre-backup.d/15-postgresql_backup.sh" "post-backup.d/10-upload-completed-status.sh" "pre-backup.d/10-upload-starting-status.sh" "pre-backup.d/11_lockfile_check.sh" "post-fail-backup.d/10-upload-fail-status.sh" "post-fail-backup.d/20-failure-notify.sh"; do
    # make sure all files are executable
    chmod +x "/etc/cloudvps-boss/${FILE}"
done

for FILE in "cloudvps-boss-progress.sh" "post-backup.d/15-clean-verbose-log.sh"; do
    if [[ -f "${FILE}" ]]; then
        # progress script is broken.
        # log cleaning should be done by logrotate.
        rm "/etc/cloudvps-boss/${FILE}"
    fi
done

for FILE in "cloudvps-boss-progress"; do
    if [[ -L "${FILE}" ]]; then
        # also remove symlink
        rm "/usr/local/bin/${FILE}"
    fi
done

if [[ -f "/usr/local/share/man/man1/duplicity.1" ]]; then
    # otherwise, /usr/bin/mandb: can't open /usr/local/share/man/man1/duplicity.1: Permission denied
    chmod 755 /usr/local/share/man/man1/duplicity.1
fi

if [[ -f "/usr/local/share/man/man1/rdiffdir.1" ]]; then
    chmod 755 /usr/local/share/man/man1/rdiffdir.1
fi

echo
lecho "If you want to receive email notifications of issues, please install"
lecho "a mailserver and add email addresses, one per line, to the following"
lecho "file: /etc/cloudvps-boss/email.conf"
echo
lecho "If you want to set up encryption, please execute the following command:"
lecho "bash /etc/cloudvps-boss/cloudvps-boss-encryption-setup.sh"
echo
lecho "CloudVPS Boss installation completed."
echo
